compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt

```diff
@@ -158,6 +179,109 @@ public sealed class TypeReference {
     }
 
     override fun isFunctionType(): Boolean = type.typeElement is KtFunctionType
+
+    override fun isNullable(): Boolean = type.typeElement is KtNullableType
+
+    private fun KtTypeReference.requireTypeName(): TypeName {
+      fun PsiElement.fail(): Nothing = throw AnvilCompilationException(
+        message = "Couldn't resolve type: $text",
+        element = this
+      )
+
+      fun KtUserType.isTypeParameter(): Boolean {
+        return parents.filterIsInstance<KtClassOrObject>().first().typeParameters.any {
+          val typeParameter = it.text.split(":").first().trim()
+          typeParameter == text
+        }
+      }
+
+      fun KtUserType.findExtendsBound(): List<FqName> {
+        return parents.filterIsInstance<KtClassOrObject>()
+          .first()
+          .typeParameters
+          .mapNotNull { it.fqName }
+      }
+
+      fun KtTypeElement.requireTypeName(): TypeName {
```

Comment:
I see that this is copy-pasta so I'm not suggesting you change anything here, but I feel like a lot of the longer methods like this are a bit difficult to follow, especially with all the different levels of nesting and could really benefit from breaking it up into more helper methods. Maybe that's something we could look at doing later.